### ISSUE TITLE
No include insertion for symbols from the main-file of the dynamic index

### ISSUE THREAD

**Issue by sam-mccall** [clangd/clangd#333]
1. edit foo.h:
   ```
   #ifndef FOO_H
   #define FOO_H
   int foo();
   #endif
   ```
2. code-complete in foo.c:
   ```
   int x = fo^
   ```

--> we see foo() completed with no include insertion
Expect to see foo() completed, and inserts `#include "foo.h"`

This appears to be bad header-guard detection when the guard is partly in the preamble. We're calling `HeaderSearch::isFileMultipleIncludeGuarded()` for the main-file, and it's returning false.

Same behavior with foo.h:
```
#pragma once
int foo();
```

however this does work (guard not in preamble):
```
int foo();
#pragma once
```

**Comment by kadircet** [comment]
so i did some digging, and this is happening because we override main file contents to only consists of preamble section while generating the PCH in https://github.com/llvm/llvm-project/blob/master/clang/lib/Frontend/PrecompiledPreamble.cpp#L332.

When serializing a PCH FileInfos in HeaderSearch is put into an OnDiskHashtable, keyed with filename + file_size + modtime. And unfortunately file size gets a different value when building a preamble from what it really is. Later on when building the AST, we load the preamble and try to look up the fileinfo for "foo.h" but lookup fails and we get a "default fileinfo struct", which doesn't know about `#pragma once`.

The dirty hack of padding the preamble section with whitespaces instead of just slicing seems to make it work. I suppose a proper fix would be to handle this in lexer similar to  code completion, i.e replacing token after preamble with EOF.

**Comment by sam-mccall** [comment]
Ah, thanks! That makes sense, and yeah either replacing with spaces (and a hilarious comment) or cutting off parsing seems reasonable.

I was thinking about whether `#ifdef` header guards were going to be hit by a second effect: the question about include-guards isn't actually well-defined for a preamble:

```
#ifndef X
#define X
int z; // preamble ends here - do we have an include guard?
#endif // we do
```

```
<reuse preamble>
int z;
#endif // uh oh, this isn't a well-formed include guard
int y; // but the HFI from the preamble named a controlling macro...
```

I'll investigate. I guess ideally we want to treat this as a header guard because it almost always is, but that seems a little unsatisfying and also maybe hard to implement.

---

And of course the question of "why do we have no test coverage" has an easy answer: some of our tests don't use preamble, and the ones that do use TestTU, which happens to use trailing-#pragma-once for header guards - the one configuration that works. Sigh, I was wondering when that was going to bite us...

**Comment by sam-mccall** [comment]
Yeah, as I feared, zeroing out the rest of the preamble file instead of truncating fixes `#pragma once` but leaves `#ifdef` broken.

EmitData in ASTWriter sees ControllingMacro=0 ControllingMacroID=0 when writing the preamble.

**Comment by sam-mccall** [comment]
So this "fixes" ifdef guards (makes them considered controlling if unclosed at end of preamble):

```
   if (PP->isRecordingPreamble() && PP->isInPrimaryFile()) {
     PP->setRecordedPreambleConditionalStack(ConditionalStack);
+    if (!ConditionalStack.empty())
+     MIOpt.ExitTopLevelConditional();
     ConditionalStack.clear();
   }
```

It's correct enough for tools, though is obviously a false positive if there's stuff in the header after the include guard. For this to cause problems you'd have to be building a PCH file from a header, which has header guards, and then including that header multiple times and relying on multiple side-effects.

Which I guess isn't totally implausible:
```
#ifndef FOO_HEADERS_LOADED
#define FOO_HEADERS_LOADED
#include <x>
#include <y>
#endif 
// tablegen-style macro calls
FOO("x", x::blah())
FOO("y", y::blah())
```

The only totally-correct thing to do I think is to record the *candidate* controlling macro and then promote it when we use the preamble. That seems like a lot of work :-(

**Comment by sam-mccall** [comment]
> Which I guess isn't totally implausible

Actually I think it's fine.

PCH files are only used for `-include` on the command-line, and the `-cc1 -include-pch` replaces the `-include` if the PCH is found, so content outside the preamble region is ignored. Actual #includes still use textual inclusion. The first one will rely on the actual `#ifdef` semantics rather than hitting the multi-header-guard code, and subsequent ones 

Probably the easiest way to summarize the correctness of this (and justify it in the code) is that the candidate header guard is valid for the preamble content itself.

**Comment by sam-mccall** [comment]
Haha, of course replacing the preamble buffer with nulls/zeros doesn't work - the idea is to match the size of the buffer we're using the preamble with, but of course that can change over time. I need to look again into the reasons for including size/mtime in the hashtable key, but I suspect zeroing them out for the main-file is the right fix.

**Comment by sam-mccall** [comment]
OK, I have a fairly ugly fix for this: for the mainfile we store the HeaderFileInfo twice, once with a size of zero. When loading a HeaderFileInfo, if a) we're loading a preamble and b) the basename of the fileentry matches the original filename stored in the preamble and c) primary lookup fails, then we retry with size 0.

But to write the testcase for this, I had to make a change that I thought was unrelated: we currently only calculate the include header for code-completion-eligible symbols, and main-file symbols aren't indexed for code completion[1]. So we also have the option of just declaring this WAI...

[1] Actually *macros* are, this inconsistency is a bug.

**Comment by sam-mccall** [comment]
Patch in https://reviews.llvm.org/D78038, not sure whether we should try to land it.

**Comment by kadircet** [comment]
> But to write the testcase for this, I had to make a change that I thought was unrelated: we currently only calculate the include header for code-completion-eligible symbols, and main-file symbols aren't indexed for code completion[1]. So we also have the option of just declaring this WAI...

this shouldn't have been happening, since we mark headers as not-mainfile-only. I believe this has happened because you changed TU.HeaderFileName rather than TU.FileName
```
bool IsMainFileOnly =
  bool IsMainFileOnly =
      SM.isWrittenInMainFile(SM.getExpansionLoc(ND->getBeginLoc())) &&
      !isHeaderFile(SM.getFileEntryForID(SM.getMainFileID())->getName(),
```

will take a look at the patch.

**Comment by jdumas** [comment]
Hi there,

I'm experiencing a similar bug with the following file structure:

### foo.h
```cpp
#pragma once

template <typename T>
struct Point {
    T x;
    T y;
};

template<typename T>
T norm(Point<T> p);

#include <foo.impl.h>
```

### foo.impl.h
```cpp
#pragma once

#include <foo.h>

template<typename T>
T norm(Point<T> p) {
    return p.x * p.x + p.y * p.y;
}
```


Specifically `clangd` is complaining when I'm in the header file `foo.h` (trying to include it twice). I tried the patch proposed above on a local build of clang on macOS and it seems to solve the issue. Is there any ETA on when it can be merged?

**Comment by jdumas** [comment]
@sam-mccall @kadircet any ETA on the patch submitted above?

**Comment by sbrocket** [comment]
+1, I have a similar situation as jdumas@ above, though a more complex case (involving generated code for a Rust/C++ FFI using https://github.com/dtolnay/cxx)

**Comment by Kaffeine** [comment]
Hello. It would be nice to have this issue fixed. [Qt Creator 8.0](https://www.qt.io/blog/qt-creator-8-released) entirely relies on clangd and (with background indexing enabled by default) the issue (same as in https://github.com/clangd/clangd/issues/333#issuecomment-630264167) is constantly popping up.

**Comment by Krzmbrzl** [comment]
I can only support that it would be really nice if this got fixed - it's quite irritating to have clangd reporting errors such as this when in fact everything compiles just fine (and is also correctly written - so we're not relying on one specific compile order or something like that).

@sam-mccall it seems that an answer of yours is awaited for in https://reviews.llvm.org/D78038 deciding whether or not your patch is to be merged.

**Comment by jdumas** [comment]
What is the status of patch [/D78038](https://reviews.llvm.org/D78038) above? Has it been moved to a GitHub PR on the LLVM project?
